File: exercise-2-refactor/exercise-2-refactor/user_service.py
# User Service — manages user accounts
# Problem: This file duplicates utility functions found in other services

import re
import uuid
from datetime import datetime
from dataclasses import dataclass


def format_date(date: datetime) -> str:
    return date.strftime("%Y-%m-%d")


def validate_email(email: str) -> bool:
    pattern = r"^[^\s@]+@[^\s@]+\.[^\s@]+$"
    return bool(re.match(pattern, email))


def format_price(cents: int) -> str:
    return f"${cents / 100:.2f}"


@dataclass
class User:
    id: str
    name: str
    email: str
    created_at: datetime
    balance: int  # in cents


def create_user(name: str, email: str) -> User:
    if not validate_email(email):
        raise ValueError(f"Invalid email: {email}")
    return User(
        id=uuid.uuid4().hex[:9],
        name=name,
        email=email,
        created_at=datetime.now(),
        balance=0,
    )


def get_user_summary(user: User) -> str:
    return (
        f"{user.name} ({user.email}) — "
        f"Joined: {format_date(user.created_at)} — "
        f"Balance: {format_price(user.balance)}"
    )

File: exercise-2-refactor/exercise-2-refactor/src/user-service.ts
// User Service — manages user accounts
// Problem: This file duplicates utility functions found in other services

function formatDate(date: Date): string {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
}

function validateEmail(email: string): boolean {
  const pattern = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return pattern.test(email);
}

function formatPrice(cents: number): string {
  return `$${(cents / 100).toFixed(2)}`;
}

export interface User {
  id: string;
  name: string;
  email: string;
  createdAt: Date;
  balance: number;
}

export function createUser(name: string, email: string): User {
  if (!validateEmail(email)) {
    throw new Error(`Invalid email: ${email}`);
  }
  return {
    id: Math.random().toString(36).substring(2, 11),
    name,
    email,
    createdAt: new Date(),
    balance: 0,
  };
}

export function getUserSummary(user: User): string {
  return `${user.name} (${user.email}) — Joined: ${formatDate(user.createdAt)} — Balance: ${formatPrice(user.balance)}`;
}
